src/enzyme_ad/jax/Implementations/StableHLOAutoDiffOpInterfaceImpl.cpp

```diff
@@ -458,6 +458,278 @@ class AutoDiffWhileFwd
+    Region *reg = op->getParentRegion();
+    op->walk([&](Operation *child) {
+      if (child == op)
+        return;
+
+      for (Value operand : child->getOperands()) {
+        if (!reg->isProperAncestor(operand.getParentRegion())) {
+          outsideRefs.insert(operand);
+        }
+      }
+    });
+  }
+
+  static LogicalResult reverseWithCheckpointing(stablehlo::WhileOp orig,
+                                                struct ReverseModeInfo revInfo,
+                                                OpBuilder &builder,
+                                                MGradientUtilsReverse *gutils,
+                                                SmallVector<Value> caches,
+                                                ArrayRef<bool> operandsActive) {
+    int64_t numIters = revInfo.info.getConstantNumIters();
+    int64_t nInner = std::sqrt(numIters);
+    int64_t nOuter = nInner;
+    assert(nInner * nOuter == numIters);
+
+    auto outer = gutils->getNewFromOriginal(orig);
+
+    SetVector<Value> outsideRefs;
+    computeOutsideRefs(orig, outsideRefs);
+
+    int numOutsideRefs = outsideRefs.size();
+    int nargs = caches.size() + 1;
+    int nrets = nargs - numOutsideRefs;
+
+    SmallVector<Value> operands;
+    for (auto [active, res] : llvm::zip(operandsActive, orig->getResults())) {
+      if (active) {
+        operands.push_back(gutils->diffe(res, builder));
+        if (!gutils->isConstantValue(res))
+          gutils->zeroDiffe(res, builder);
+      }
+    }
+
+    OpBuilder::InsertionGuard guard(builder);
+
+    stablehlo::WhileOp revOuter =
+        makeForLoop(builder, orig.getLoc(), 0, nOuter, 1, operands);
+
+    Block *revOuterBody = &revOuter.getBody().front();
+    builder.setInsertionPointToStart(revOuterBody);
+
+    Value outerStep = builder.create<stablehlo::SubtractOp>(
+        orig.getLoc(), makeI64Constant(orig.getLoc(), builder, nOuter - 1),
+        revOuterBody->getArgument(0));
+    Value outerStart = builder.create<stablehlo::MulOp>(
+        orig.getLoc(), makeI64Constant(orig.getLoc(), builder, nInner),
+        outerStep);
+
+    Value lastCache = nullptr;
+
+    SmallVector<Value> cacheVals(nargs - 1, nullptr);
+    for (int i = 0; i < nrets - 1; ++i)
+      lastCache = cacheVals[i] = gutils->popCache(caches[i], builder);
+
+    builder.setInsertionPoint(revOuter);
+
+    IRMapping mapping;
+
+    for (int i = nrets - 1, refIdx = 0; i < nargs - 1; ++i) {
+      Value cached = cacheVals[i] = gutils->popCache(caches[i], builder);
+      mapping.map(outsideRefs[refIdx], cached);
+      refIdx++;
+    }
+
+    if (lastCache)
+      builder.setInsertionPointAfterValue(lastCache);
+    else
+      builder.setInsertionPointToStart(revOuterBody);
+
+    operands.assign(revOuterBody->getArguments().slice(1).begin(),
+                    revOuterBody->getArguments().slice(1).end());
+
+    auto revInner = makeForLoop(builder, orig.getLoc(), 0, nInner, 1, operands);
```

Comment:
we should just clone all the attributes of the original loop -- excludingf the checkpointing attr